activerecord/lib/active_record/connection_adapters/abstract/transaction.rb

```diff
@@ -109,13 +109,15 @@ def dirty?
         @dirty
       end
 
-      def add_record(record, ensure_finalize = true)
+      # Add the record to the current transaction, used by after_commit (etc) callbacks.
+      # Returns true if the record was added to the transaction, false if it had been previously added.
+      def uniquely_add_record(record, ensure_finalize = true)
         @records ||= []
         if ensure_finalize
-          @records << record
+          @records << record unless @records.map(&:__id__).include?(record.__id__)
```

Comment:
This uses `__id__` for the same reason as https://github.com/rails/rails/pull/38990

Reply:
Instead of keeping it as an array we then have to scan (and assigning object IDs in the process), let's switch `@records` to be a `Set.new.compare_by_identity`. Then we can just:

```suggestion
          @records.add?(record)
```

(and as the set absolutely guarantees uniqueness, we can drop the `.uniq` calls elsewhere)